processor/src/host/debug.rs

```diff
@@ -1,52 +1,193 @@
+        };
+
+        print_interval(mem_interval);
+    }
+
+    /// Prints locals in provided interval.
+    ///
+    /// If `print_all` is true, the interval should contain all indexes of locals available for the
+    /// current procedure.
+    fn print_local_interval<S: ProcessState>(
+        &self,
+        process: &S,
+        interval: (u32, u32),
+        num_locals: u32,
+        print_all: bool,
+    ) {
+        let mut local_mem_interval = Vec::new();
+        let local_memory_offset = self.fmp - num_locals + 1;
+        for index in interval.0..interval.1 + 1 {
+            local_mem_interval
+                .push((index, process.get_mem_value(self.ctx, index + local_memory_offset)))
+        }
+
+        if print_all {
+            println!("Local state before step {} for the context {}:", self.clk, self.ctx)
+        } else if interval.0 == interval.1 {
+            println!(
+                "Local state before step {} for the context {} at index {}:",
+                self.clk, self.ctx, interval.0
+            )
+        } else {
+            println!(
+                "Local state before step {} for the context {} in the interval [{}, {}]:",
+                self.clk, self.ctx, interval.0, interval.1
+            )
+        };
+
+        print_interval(local_mem_interval);
+    }
+}
 
-    // print all items except for the last one
-    println!("Stack state before step {clk}:");
-    for (i, element) in stack.iter().take(num_items - 1).enumerate() {
-        println!("├── {i:>2}: {element}");
+fn print_interval(mem_interval: Vec<(u32, Option<Word>)>) {
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem_interval.iter().take(mem_interval.len() - 1) {
+        if let Some(value) = value {
+            print_word(*addr, value, false);
+        } else {
+            println!("├── {addr:.<10}: EMPTY");
+        }
     }
 
-    // print the last item, and in case the stack has more items, print the total number of
-    // un-printed items
-    let i = num_items - 1;
-    if num_items == stack.len() {
-        println!("└── {i:>2}: {}", stack[i]);
+    // print the last memory value
+    if let Some((addr, value)) = mem_interval.last() {
+        if let Some(value) = value {
+            print_word(*addr, value, true);
+        } else {
+            println!("└── {addr:.<10}: EMPTY\n");
+        }
     } else {
-        println!("├── {i:>2}: {}", stack[i]);
-        println!("└── ({} more items)", stack.len() - num_items);
+        println!("└── EMPTY\n");
     }
 }
 
-#[cfg(not(feature = "std"))]
-fn print_vm_stack(_clk: u32, _stack: Vec<Felt>, _n: usize) {
-    // in no_std environments, this is a NOOP
+fn print_word(addr: u32, value: &Word, last: bool) {
+    if last {
+        print!("└── ");
+    } else {
+        print!("├── ");
+    }
+    println!("{addr:.<10}: [{}, {}, {}, {}]", value[0], value[1], value[2], value[3],);
+    if last {
+        println!();
+    }
 }
```

Comment:
I would replace this with:
```Rust
fn print_mem_address(addr: u32, value: Option<Word>, is_last: bool, is_local: bool) {

}
```
And handle all variations internally.

For memory address, you can use the following format `{add:#010x}`. For, local, you can use `{addr:>5}`.

Also, we should probably impose some width formatting for the values. How many characters does u64 require?

Reply:
u64 require 20 characters, but spacing each value with 20 white spaces will look strange. Here I can try to use the padding which depends on the maximum length of the value. In that case width of the output values will be equal, but potentially smaller than 20.

Reply:
Yep - I think that's a good idea!